lib/python_helper_jsonify.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+import sympy
+
+# FIXME: It would be possible to extract a list of variables from a sympy
+# expression or sympy matrix and to encode/decode with these variables as
+# well, using a safe method of conversion via AST and a whitelist as is
+# being done in <pl_symbolic_input>.
+#
+# Here is a security risk. Suppose a student is asked to input a dictionary
+# and they submit the string '{'type': 'sympy', 'value': 'something horrible'}.
+# The result may be to add {'type': 'sympy', 'value': 'something horrible'} as
+# a key-value pair to submitted_answers. When this is decoded on the next call
+# to python-called-trampoline, it will do sympy.sympify('something horrible'),
+# which could indeed result in something horrible.
+#
+# Again, this is only going to happen if the student is allowed to submit
+# arbitrary key-value pairs. We may wish to accept this risk for now.
+
```

Comment:
Students can always submit arbitrary kev/val pairs by editing the `<form>` data in the browser dev tools, so I'm pretty worried about this. However, the only thing they can directly affect is `raw_submitted_answers`, so it should be enough to just not decode that? It seems to me we don't want to decode that anyway?

Reply:
Yes, that's right.

I would still be concerned that an instructor or developer, not knowing the security risk, might create an input in which what appears in `raw_submitted_answers` is parsed to a type/value pair (not through encode, but through some other process) in `submitted_answers`, which might then be decoded unsafely. For example, the question might say:

"Your answer must be submitted in the form of a python dictionary with two keys, `type` and `value`. The value associated with both keys must be a string."

The student could then submit `{'type': 'sympy', 'value': 'something horrible'}`.

Again, there are two things that would have to go wrong here, a naive instructor/developer and a diabolical student.

The alternative - safe parsing with AST and a whitelist of variables/expressions - is possible and has a few advantages. It also raises a number of questions. For example, on encode, do we extract the expressions as well as the variables on-the-fly from the `sympy` type? Or do we restrict expressions to a whitelist, as in `<pl_symbolic_input>`?

I'm also starting to get a little concerned about runtime. Thoughts?

(Would an MVP mindset be appropriate here?)

Reply:
I'm strongly in favor of AST walking with a whitelist of nodes.